python/tvm/relay/frontend/common.py

```diff
@@ -474,6 +476,50 @@ def infer_channels(inputs, transpose=False):
     return channels
 
 
+def infer_value(input_val, params):
+    """A hack for getting the value of an expression by evaluating a
+    portion of the relay graph. This is often needed for functions that
+    whose output shape depends on the value of a tensor.
+    """
+    from tvm.contrib import graph_runtime
+    # Check that all free variables have associated parameters.
+    assert all(var.name_hint in params.keys() for var in analysis.free_vars(
+        input_val)), "All inputs to infer must be available in params."
+    func = _expr.Function(analysis.free_vars(input_val), input_val)
+    with tvm.relay.build_config(opt_level=0):
+        graph, lib, params = tvm.relay.build(func, target="llvm", params=params)
+    ctx = tvm.context("llvm", 0)
+    m = graph_runtime.create(graph, lib, ctx)
+    m.set_input(**params)
+    m.run()
+    return m.get_output(0)
+
+
+def infer_value_simulated(input_val, params):
+    """Extention to infer_value that can be used when some input
+    values are missing. This function creats dummy inputs with the same
+    shape and random values then calls infer_value. This is helpful when
+    implementing certain onnx operators where we need to evaluate the graph
+    to determine a static shape.
+    """
+    fake_params = []
+    # Add a fake copy of all missing params.
+    for free_param in analysis.free_vars(input_val):
+        if free_param.name_hint not in params:
+            fp_dtype = free_param.type_annotation.dtype
+            fp_shape = [s.value for s in free_param.type_annotation.shape]
+            fake_params.append(free_param)
+            params[free_param.name_hint] = tvm.nd.array(
+                np.random.rand(*fp_shape).astype(fp_dtype)
+            )
+    # Now infer the value.
+    output_value = infer_value(input_val, params)
```

Comment:
This is a very useful function that I think will come in handy for other frontends. How will it work with dynamic shapes? For example, if my model's input shape is `(Any, 12, 128)`, `infer_value_simulated` will only be able to infer a value for the real shape that you passed in.

Reply:
Yeah, `infer_value_simulated` is essentially a work around for dynamic shapes in frameworks since relay currently requires static shaping. I'm using it here to evaluate the graph to a fixed shape based on the inputs. Down the line I could imagine relay having better support for dynamic shaped arguments.

Reply:
I think a goal is to support what we can today and start to incrementally generalize the ops to be more dynamic. There is tension between more static ops (which can generate better code) vs. fully dynamic ones which are convenient. My 2c is that we should probably have two variants and a set of passes which try to detect and lower the more dynamic variants into the static ones where applicable. We have already done a bit of this, for example evaluating shape_of to a constant when the shape is static.

Reply:
I have a question, what if the value is data dependent?

Reply:
Although it's possible that the values for shapes are based on the inputs it's very uncommon. However, if this were the case there's a high chance the inferred shape would be invalid and cause an error since the simulated input is just random noise.